src/Microsoft.ML.Data/DataLoadSave/Binary/Codecs.cs

```diff
@@ -1276,7 +1272,112 @@ private bool GetKeyCodec(Stream definitionStream, out IValueCodec codec)
                 Contracts.CheckDecode((ulong)count <= itemType.RawKind.ToMaxInt());
                 Contracts.CheckDecode(contiguous || count == 0);
 
-                type = new KeyType(itemType.RawKind, min, count, contiguous);
+                type = new KeyType(itemType.RawType, count);
```

Comment:
>KeyType [](start = 27, length = 7)

I'd feel better with some tests of this. Maybe, *in the old version*, prepare a completely empty binary IDV file (empty in the sense of no rows), but with a schema where you exercise all these strange conditions. You check in that IDV file. (It's a binary file, but a relatively small one.)

So, for instance, I might create something that looks like this:

    dotnet MML.dll savedata loader=text{col=A:U4[0-2]:0 col=B:U4[2-5]:0 col=C:U8[5-10]:0 col=D:U2[0-*]:0} dout=bubba.idv

If I were subsequently to do this:

    dotnet MML.dll showschema data=bubba.idv

Then what I see is this:

```
4 columns:
  A: Key<U4, 0-2>
  B: Key<U4, 2-5>
  C: Key<U8, 5-10>
  D: Key<U2, 0-*>
```

whereas what I might see *after* your changes is something more like this:

```
4 columns:
  A: Key<U4, 0-2>
  B: Key<U4, 0-3>
  C: Key<U8, 0-5>
  D: Key<U2, 0-65534>
```

But... like I say might be nice to test it. (I'm doing this through command line, but that's just for demonstration purposes. The test itself should probably be precise and programmatic.)

Reply:
BTW probably have to be more comprehensive than I was, that was just a small example.

---
In reply to: [249132474](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249132474) [](ancestors = 249132474)

Reply:
I agree with adding a test, and good catch I had forgotten about that!! 

---
In reply to: [249132474](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249132474) [](ancestors = 249132474)